lib/api/objectPutPart.js

```diff
@@ -163,6 +167,39 @@ export default function objectPutPart(authInfo, request, streamingV4Params, log,
                     return next();
                 });
         },
+        // Get any pre-existing part.
+        next => {
+            const partKey = `${uploadId}${splitter}${paddedPartNumber}`;
+            return metadata.getObjectMD(mpuBucketName, partKey, log,
+                (err, res) => {
+                    // If there is no object with the same key, continue.
+                    if (err && !err.NoSuchKey) {
+                        log.error('error getting current part (if any)', {
+                            error: err,
+                            method: 'objectPutPart::metadata.getObjectMD',
+                        });
+                        return next(err);
+                    }
+                    return next(null, res);
+                });
+        },
+        // If there is a pre-existing part, update utapi metrics and delete the
+        // pre-existing part to prevent orphaned data.
+        (res, next) => {
+            if (res) {
+                prevObjectSize = res['content-length'];
+                // Pull locations to clean up any potential orphans in data if
+                // object put is an overwrite of a pre-existing object with the
+                // same key and part number.
+                const oldLocations = Array.isArray(res.partLocations) ?
+                    res.partLocations : [res.partLocations];
+                log.trace('Overwriting MPU part, deleting data');
+                data.batchDelete(oldLocations, logger
```

Comment:
we should only do the delete after the other part has been successfully stored in data and metadata.